[FEATURE] Configurable multi-threaded checkpoint speedup

jdimatteo opened the pull request:
Add configuration option to speed up execution with mulithreading, informed by [profiling and benchmark results]( https://github.com/great-expectations/great_expectations/pull/3287).

Compared to the [test_taxi_trips_benchmark[100-False-V2]](https://github.com/great-expectations/great_expectations/blob/857c4835bd0b0dbad4a8d0c7dbc48aeb0868f921/tests/performance/test_bigquery_benchmarks.py#L25) fastest single threaded [baseline time of 5 minutes 44.7 seconds](https://github.com/great-expectations/great_expectations/blob/857c4835bd0b0dbad4a8d0c7dbc48aeb0868f921/tests/performance/results/initial_baseline_run_3.json#L1000), this implementation's [multithreaded run time is 22.5 seconds](https://github.com/great-expectations/great_expectations/blob/857c4835bd0b0dbad4a8d0c7dbc48aeb0868f921/tests/performance/results/minimal_multithreading_run_3.json#L1000) -- this is a speedup of **[15.3x faster](https://docs.google.com/spreadsheets/d/1O9tQEGZS4IYZJWjWUoK_FYkJqNVRkycTZYp2SoXdRS4/edit#gid=0&range=J7)**.

[![Runtime Chart showing Scaling](https://docs.google.com/spreadsheets/d/e/2PACX-1vRLoppdlbSYy87cidbfDixllHFpa0RbjNS-xg5uZBHECIGdPOdcTC1CK-f8zHp4eH5a40VFq1GTzmKE/pubchart?oid=1237597596&format=image)](https://docs.google.com/spreadsheets/d/1O9tQEGZS4IYZJWjWUoK_FYkJqNVRkycTZYp2SoXdRS4/edit#gid=0)
* Above chart data is available at [this google sheet](https://docs.google.com/spreadsheets/d/1O9tQEGZS4IYZJWjWUoK_FYkJqNVRkycTZYp2SoXdRS4/edit?usp=sharing).
* Benchmark results can be compared with `py.test-benchmark compare` as shown [here](https://gist.githubusercontent.com/jdimatteo/9087bac5aa8981b39f858b3a05f197e8/raw/213bc2b7e23ed397cdbbc949f02032aa0cd06153/gistfile1.txt).

With mulithreading, data docs rendering becomes the bottleneck, and additional optimization could be done there for further speedups. I would *greatly appreciate anyone submitting PRs speeding up data docs rendering* -- please reach out to me at jdimatteo@gmail.com if you would like to collaborate on this :pray:.  When data docs rendering time is included, benchmark speedup is only **6.3x faster** (for `test_taxi_trips_benchmark[100-True-V2]` [baseline is 6 minutes 24.1 seconds](https://github.com/great-expectations/great_expectations/blob/857c4835bd0b0dbad4a8d0c7dbc48aeb0868f921/tests/performance/results/initial_baseline_run_2.json#L1084) and [multithreaded is 60.8 seconds](https://github.com/great-expectations/great_expectations/blob/857c4835bd0b0dbad4a8d0c7dbc48aeb0868f921/tests/performance/results/minimal_multithreading_run_2.json#L1084)).

This multithreading implementation and configuration option is explicitly at [maturity level "experimental"](https://docs.greatexpectations.io/docs/contributing/contributing_maturity) because more testing is needed for various backends, more user facing documentation is needed, and so that we don't yet have to commit to the API added here being stable. Because multithreading is disabled by default, it is safe to merge at this time.

Changes proposed in this pull request:
1. Adds optional `ConcurrencyConfig` to `DataContextConfig`, defaulting to concurrency disabled.
2. If enabled in `ConcurrencyConfig`, executes `Checkpoint.run` with multiple threads to speed up V3 Batch Request API execution.
3. If enabled in `ConcurrencyConfig`, executes `ActionListValidationOperator.run` with multiple threads to speed up V2 Batch Kwargs API execution.
4. Added classes `AsyncExecutor` and `AsyncFuture` to encapsulate multithreaded asynchronous execution logic, to facilitate a single code path (with no code duplication) for both when concurrency is enabled and disabled.
5. Adds full benchmark results for baseline: `tests/performance/results/initial_baseline_*.json` using develop commit 3ea9c8e78334b55dc29a87b051dd2fc69d01acad).
6. Adds full performance benchmark results using this multithreaded implementation (and no other changes from 3ea9c8e78334b55dc29a87b051dd2fc69d01acad to ensure a direct comparison with the baseline): `tests/performance/results/minimal_multithreading_*.json`.
7. Enhances `tests/performance/test_bigquery_benchmarks.py` to also test performance using the V2 Batch Kwargs API.
8. Enhances `tests/performance/test_bigquery_benchmarks.py` to optionally save Data Docs HTML to the environment variable `GE_BENCHMARK_HTML_DIRECTORY` to ease testing/debugging.
9. Make benchmark tests more robust to implementation changes (i.e. less fragile), fixing the current `develop` branch `HEAD` [test failures](https://dev.azure.com/great-expectations/great_expectations/_build/results?buildId=16404&view=logs&j=96801852-5ae4-523f-8df1-841bda737170&t=27884d75-315f-57e5-0d6f-422e5ea1a6a4&l=151).
10. To remove concurrency bottlenecks, configure SqlAlchemy with `create_engine` parameters [pool_size](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.pool_size) and [max_overflow](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.max_overflow). Also optionally patch urllib3 to enable a higher default max pool size to reduce concurrency bottlenecks.

Previous Design Review notes:
- [Meeting notes](https://docs.google.com/document/d/1-Lik3Sk3zLGwjP7bVxpI72VmWAO17BGcwS2RxeYHhg8/edit) between @talagluck, @kconvey, @jcampbell, @Shinnnyshinshin, and @jdimatteo.
- Discussion at https://github.com/great-expectations/great_expectations/pull/3287
- Discussion at https://github.com/great-expectations/great_expectations/pull/3280


### Definition of Done

- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

jdimatteo committed:
Added test case for V2 API

Also made test less fragile so passes for V3 even when new keys are added.

jdimatteo committed:
Optionally write HTML to a specified non-tmp directory.

jdimatteo committed:
Add script to run performance benchmark multiple times.

jdimatteo committed:
Add full performance baseline results (replacing the prior incomplete baseline result).

```
$ ./tests/performance/run_benchmark_multiple_times.sh initial_baseline
Sun 05 Sep 2021 12:30:05 PM MDT
+ pytest tests/performance/test_bigquery_benchmarks.py --benchmark-json=tests/performance/results/initial_baseline_run_1.json --no-spark --no-postgresql --bigquery --performance-tests -q -p no:warnings
........................                                                                                                                                                                                                                                                                                               [100%]

-------------------------------------------------------------------------------------------- benchmark: 24 tests ---------------------------------------------------------------------------------------------
Name (time in s)                                 Min                 Max                Mean            StdDev              Median               IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_taxi_trips_benchmark[1-True-V3]          4.2233 (1.0)        4.2233 (1.0)        4.2233 (1.0)      0.0000 (1.0)        4.2233 (1.0)      0.0000 (1.0)           0;0  0.2368 (1.0)           1           1
test_taxi_trips_benchmark[1-False-V3]         4.3585 (1.03)       4.3585 (1.03)       4.3585 (1.03)     0.0000 (1.0)        4.3585 (1.03)     0.0000 (1.0)           0;0  0.2294 (0.97)          1           1
test_taxi_trips_benchmark[1-True-V2]          4.7968 (1.14)       4.7968 (1.14)       4.7968 (1.14)     0.0000 (1.0)        4.7968 (1.14)     0.0000 (1.0)           0;0  0.2085 (0.88)          1           1
test_taxi_trips_benchmark[1-False-V2]         5.2558 (1.24)       5.2558 (1.24)       5.2558 (1.24)     0.0000 (1.0)        5.2558 (1.24)     0.0000 (1.0)           0;0  0.1903 (0.80)          1           1
test_taxi_trips_benchmark[2-False-V3]         8.3849 (1.99)       8.3849 (1.99)       8.3849 (1.99)     0.0000 (1.0)        8.3849 (1.99)     0.0000 (1.0)           0;0  0.1193 (0.50)          1           1
test_taxi_trips_benchmark[2-True-V3]          8.4396 (2.00)       8.4396 (2.00)       8.4396 (2.00)     0.0000 (1.0)        8.4396 (2.00)     0.0000 (1.0)           0;0  0.1185 (0.50)          1           1
test_taxi_trips_benchmark[2-False-V2]         8.7447 (2.07)       8.7447 (2.07)       8.7447 (2.07)     0.0000 (1.0)        8.7447 (2.07)     0.0000 (1.0)           0;0  0.1144 (0.48)          1           1
test_taxi_trips_benchmark[2-True-V2]          8.8481 (2.10)       8.8481 (2.10)       8.8481 (2.10)     0.0000 (1.0)        8.8481 (2.10)     0.0000 (1.0)           0;0  0.1130 (0.48)          1           1
test_taxi_trips_benchmark[4-True-V3]         15.4317 (3.65)      15.4317 (3.65)      15.4317 (3.65)     0.0000 (1.0)       15.4317 (3.65)     0.0000 (1.0)           0;0  0.0648 (0.27)          1           1
test_taxi_trips_benchmark[4-True-V2]         16.2528 (3.85)      16.2528 (3.85)      16.2528 (3.85)     0.0000 (1.0)       16.2528 (3.85)     0.0000 (1.0)           0;0  0.0615 (0.26)          1           1
test_taxi_trips_benchmark[4-False-V3]        17.2043 (4.07)      17.2043 (4.07)      17.2043 (4.07)     0.0000 (1.0)       17.2043 (4.07)     0.0000 (1.0)           0;0  0.0581 (0.25)          1           1
test_taxi_trips_benchmark[4-False-V2]        18.1744 (4.30)      18.1744 (4.30)      18.1744 (4.30)     0.0000 (1.0)       18.1744 (4.30)     0.0000 (1.0)           0;0  0.0550 (0.23)          1           1
test_taxi_trips_benchmark[8-True-V3]         29.2475 (6.93)      29.2475 (6.93)      29.2475 (6.93)     0.0000 (1.0)       29.2475 (6.93)     0.0000 (1.0)           0;0  0.0342 (0.14)          1           1
test_taxi_trips_benchmark[8-False-V3]        29.9335 (7.09)      29.9335 (7.09)      29.9335 (7.09)     0.0000 (1.0)       29.9335 (7.09)     0.0000 (1.0)           0;0  0.0334 (0.14)          1           1
test_taxi_trips_benchmark[8-True-V2]         33.0360 (7.82)      33.0360 (7.82)      33.0360 (7.82)     0.0000 (1.0)       33.0360 (7.82)     0.0000 (1.0)           0;0  0.0303 (0.13)          1           1
test_taxi_trips_benchmark[8-False-V2]        36.7383 (8.70)      36.7383 (8.70)      36.7383 (8.70)     0.0000 (1.0)       36.7383 (8.70)     0.0000 (1.0)           0;0  0.0272 (0.11)          1           1
test_taxi_trips_benchmark[16-False-V3]       61.3593 (14.53)     61.3593 (14.53)     61.3593 (14.53)    0.0000 (1.0)       61.3593 (14.53)    0.0000 (1.0)           0;0  0.0163 (0.07)          1           1
test_taxi_trips_benchmark[16-True-V3]        61.6513 (14.60)     61.6513 (14.60)     61.6513 (14.60)    0.0000 (1.0)       61.6513 (14.60)    0.0000 (1.0)           0;0  0.0162 (0.07)          1           1
test_taxi_trips_benchmark[16-True-V2]        62.8486 (14.88)     62.8486 (14.88)     62.8486 (14.88)    0.0000 (1.0)       62.8486 (14.88)    0.0000 (1.0)           0;0  0.0159 (0.07)          1           1
test_taxi_trips_benchmark[16-False-V2]       68.7432 (16.28)     68.7432 (16.28)     68.7432 (16.28)    0.0000 (1.0)       68.7432 (16.28)    0.0000 (1.0)           0;0  0.0145 (0.06)          1           1
test_taxi_trips_benchmark[100-True-V2]      384.9857 (91.16)    384.9857 (91.16)    384.9857 (91.16)    0.0000 (1.0)      384.9857 (91.16)    0.0000 (1.0)           0;0  0.0026 (0.01)          1           1
test_taxi_trips_benchmark[100-False-V3]     391.8299 (92.78)    391.8299 (92.78)    391.8299 (92.78)    0.0000 (1.0)      391.8299 (92.78)    0.0000 (1.0)           0;0  0.0026 (0.01)          1           1
test_taxi_trips_benchmark[100-False-V2]     435.7602 (103.18)   435.7602 (103.18)   435.7602 (103.18)   0.0000 (1.0)      435.7602 (103.18)   0.0000 (1.0)           0;0  0.0023 (0.01)          1           1
test_taxi_trips_benchmark[100-True-V3]      477.5333 (113.07)   477.5333 (113.07)   477.5333 (113.07)   0.0000 (1.0)      477.5333 (113.07)   0.0000 (1.0)           0;0  0.0021 (0.01)          1           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
24 passed in 2225.06s (0:37:05)

real    37m10.285s
user    3m10.884s
sys     0m15.760s
+ set +x
Sun 05 Sep 2021 01:07:15 PM MDT
+ pytest tests/performance/test_bigquery_benchmarks.py --benchmark-json=tests/performance/results/initial_baseline_run_2.json --no-spark --no-postgresql --bigquery --performance-tests -q -p no:warnings
........................                                                                                                                                                                                                                                                                                               [100%]

-------------------------------------------------------------------------------------------- benchmark: 24 tests ---------------------------------------------------------------------------------------------
Name (time in s)                                 Min                 Max                Mean            StdDev              Median               IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_taxi_trips_benchmark[1-False-V3]         3.3227 (1.0)        3.3227 (1.0)        3.3227 (1.0)      0.0000 (1.0)        3.3227 (1.0)      0.0000 (1.0)           0;0  0.3010 (1.0)           1           1
test_taxi_trips_benchmark[1-True-V3]          3.5200 (1.06)       3.5200 (1.06)       3.5200 (1.06)     0.0000 (1.0)        3.5200 (1.06)     0.0000 (1.0)           0;0  0.2841 (0.94)          1           1
test_taxi_trips_benchmark[1-False-V2]         3.5963 (1.08)       3.5963 (1.08)       3.5963 (1.08)     0.0000 (1.0)        3.5963 (1.08)     0.0000 (1.0)           0;0  0.2781 (0.92)          1           1
test_taxi_trips_benchmark[1-True-V2]          3.9647 (1.19)       3.9647 (1.19)       3.9647 (1.19)     0.0000 (1.0)        3.9647 (1.19)     0.0000 (1.0)           0;0  0.2522 (0.84)          1           1
test_taxi_trips_benchmark[2-False-V3]         6.9249 (2.08)       6.9249 (2.08)       6.9249 (2.08)     0.0000 (1.0)        6.9249 (2.08)     0.0000 (1.0)           0;0  0.1444 (0.48)          1           1
test_taxi_trips_benchmark[2-True-V3]          7.0319 (2.12)       7.0319 (2.12)       7.0319 (2.12)     0.0000 (1.0)        7.0319 (2.12)     0.0000 (1.0)           0;0  0.1422 (0.47)          1           1
test_taxi_trips_benchmark[2-False-V2]         7.0574 (2.12)       7.0574 (2.12)       7.0574 (2.12)     0.0000 (1.0)        7.0574 (2.12)     0.0000 (1.0)           0;0  0.1417 (0.47)          1           1
test_taxi_trips_benchmark[2-True-V2]          7.7079 (2.32)       7.7079 (2.32)       7.7079 (2.32)     0.0000 (1.0)        7.7079 (2.32)     0.0000 (1.0)           0;0  0.1297 (0.43)          1           1
test_taxi_trips_benchmark[4-False-V3]        13.5262 (4.07)      13.5262 (4.07)      13.5262 (4.07)     0.0000 (1.0)       13.5262 (4.07)     0.0000 (1.0)           0;0  0.0739 (0.25)          1           1
test_taxi_trips_benchmark[4-False-V2]        13.7150 (4.13)      13.7150 (4.13)      13.7150 (4.13)     0.0000 (1.0)       13.7150 (4.13)     0.0000 (1.0)           0;0  0.0729 (0.24)          1           1
test_taxi_trips_benchmark[4-True-V3]         15.3856 (4.63)      15.3856 (4.63)      15.3856 (4.63)     0.0000 (1.0)       15.3856 (4.63)     0.0000 (1.0)           0;0  0.0650 (0.22)          1           1
test_taxi_trips_benchmark[4-True-V2]         16.5584 (4.98)      16.5584 (4.98)      16.5584 (4.98)     0.0000 (1.0)       16.5584 (4.98)     0.0000 (1.0)           0;0  0.0604 (0.20)          1           1
test_taxi_trips_benchmark[8-False-V3]        28.7145 (8.64)      28.7145 (8.64)      28.7145 (8.64)     0.0000 (1.0)       28.7145 (8.64)     0.0000 (1.0)           0;0  0.0348 (0.12)          1           1
test_taxi_trips_benchmark[8-False-V2]        29.4374 (8.86)      29.4374 (8.86)      29.4374 (8.86)     0.0000 (1.0)       29.4374 (8.86)     0.0000 (1.0)           0;0  0.0340 (0.11)          1           1
test_taxi_trips_benchmark[8-True-V2]         30.3670 (9.14)      30.3670 (9.14)      30.3670 (9.14)     0.0000 (1.0)       30.3670 (9.14)     0.0000 (1.0)           0;0  0.0329 (0.11)          1           1
test_taxi_trips_benchmark[8-True-V3]         30.9665 (9.32)      30.9665 (9.32)      30.9665 (9.32)     0.0000 (1.0)       30.9665 (9.32)     0.0000 (1.0)           0;0  0.0323 (0.11)          1           1
test_taxi_trips_benchmark[16-False-V3]       53.7829 (16.19)     53.7829 (16.19)     53.7829 (16.19)    0.0000 (1.0)       53.7829 (16.19)    0.0000 (1.0)           0;0  0.0186 (0.06)          1           1
test_taxi_trips_benchmark[16-False-V2]       56.1152 (16.89)     56.1152 (16.89)     56.1152 (16.89)    0.0000 (1.0)       56.1152 (16.89)    0.0000 (1.0)           0;0  0.0178 (0.06)          1           1
test_taxi_trips_benchmark[16-True-V2]        60.8217 (18.31)     60.8217 (18.31)     60.8217 (18.31)    0.0000 (1.0)       60.8217 (18.31)    0.0000 (1.0)           0;0  0.0164 (0.05)          1           1
test_taxi_trips_benchmark[16-True-V3]        69.8583 (21.02)     69.8583 (21.02)     69.8583 (21.02)    0.0000 (1.0)       69.8583 (21.02)    0.0000 (1.0)           0;0  0.0143 (0.05)          1           1
test_taxi_trips_benchmark[100-False-V3]     328.5219 (98.87)    328.5219 (98.87)    328.5219 (98.87)    0.0000 (1.0)      328.5219 (98.87)    0.0000 (1.0)           0;0  0.0030 (0.01)          1           1
test_taxi_trips_benchmark[100-False-V2]     348.5321 (104.90)   348.5321 (104.90)   348.5321 (104.90)   0.0000 (1.0)      348.5321 (104.90)   0.0000 (1.0)           0;0  0.0029 (0.01)          1           1
test_taxi_trips_benchmark[100-True-V3]      382.3974 (115.09)   382.3974 (115.09)   382.3974 (115.09)   0.0000 (1.0)      382.3974 (115.09)   0.0000 (1.0)           0;0  0.0026 (0.01)          1           1
test_taxi_trips_benchmark[100-True-V2]      384.1241 (115.61)   384.1241 (115.61)   384.1241 (115.61)   0.0000 (1.0)      384.1241 (115.61)   0.0000 (1.0)           0;0  0.0026 (0.01)          1           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
24 passed in 1938.12s (0:32:18)

real    32m23.723s
user    3m14.527s
sys     0m16.380s
+ set +x
Sun 05 Sep 2021 01:39:39 PM MDT
+ pytest tests/performance/test_bigquery_benchmarks.py --benchmark-json=tests/performance/results/initial_baseline_run_3.json --no-spark --no-postgresql --bigquery --performance-tests -q -p no:warnings
........................                                                                                                                                                                                                                                                                                               [100%]

-------------------------------------------------------------------------------------------- benchmark: 24 tests ---------------------------------------------------------------------------------------------
Name (time in s)                                 Min                 Max                Mean            StdDev              Median               IQR            Outliers     OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_taxi_trips_benchmark[1-True-V3]          3.2818 (1.0)        3.2818 (1.0)        3.2818 (1.0)      0.0000 (1.0)        3.2818 (1.0)      0.0000 (1.0)           0;0  0.3047 (1.0)           1           1
test_taxi_trips_benchmark[1-False-V3]         3.8316 (1.17)       3.8316 (1.17)       3.8316 (1.17)     0.0000 (1.0)        3.8316 (1.17)     0.0000 (1.0)           0;0  0.2610 (0.86)          1           1
test_taxi_trips_benchmark[1-False-V2]         3.8848 (1.18)       3.8848 (1.18)       3.8848 (1.18)     0.0000 (1.0)        3.8848 (1.18)     0.0000 (1.0)           0;0  0.2574 (0.84)          1           1
test_taxi_trips_benchmark[1-True-V2]          4.4692 (1.36)       4.4692 (1.36)       4.4692 (1.36)     0.0000 (1.0)        4.4692 (1.36)     0.0000 (1.0)           0;0  0.2238 (0.73)          1           1
test_taxi_trips_benchmark[2-False-V2]         7.2650 (2.21)       7.2650 (2.21)       7.2650 (2.21)     0.0000 (1.0)        7.2650 (2.21)     0.0000 (1.0)           0;0  0.1376 (0.45)          1           1
test_taxi_trips_benchmark[2-False-V3]         7.8152 (2.38)       7.8152 (2.38)       7.8152 (2.38)     0.0000 (1.0)        7.8152 (2.38)     0.0000 (1.0)           0;0  0.1280 (0.42)          1           1
test_taxi_trips_benchmark[2-True-V2]          8.2846 (2.52)       8.2846 (2.52)       8.2846 (2.52)     0.0000 (1.0)        8.2846 (2.52)     0.0000 (1.0)           0;0  0.1207 (0.40)          1           1
test_taxi_trips_benchmark[2-True-V3]          8.4899 (2.59)       8.4899 (2.59)       8.4899 (2.59)     0.0000 (1.0)        8.4899 (2.59)     0.0000 (1.0)           0;0  0.1178 (0.39)          1           1
test_taxi_trips_benchmark[4-False-V3]        13.0151 (3.97)      13.0151 (3.97)      13.0151 (3.97)     0.0000 (1.0)       13.0151 (3.97)     0.0000 (1.0)           0;0  0.0768 (0.25)          1           1
test_taxi_trips_benchmark[4-True-V3]         13.7633 (4.19)      13.7633 (4.19)      13.7633 (4.19)     0.0000 (1.0)       13.7633 (4.19)     0.0000 (1.0)           0;0  0.0727 (0.24)          1           1
test_taxi_trips_benchmark[4-False-V2]        14.2186 (4.33)      14.2186 (4.33)      14.2186 (4.33)     0.0000 (1.0)       14.2186 (4.33)     0.0000 (1.0)           0;0  0.0703 (0.23)          1           1
test_taxi_trips_benchmark[4-True-V2]         15.4466 (4.71)      15.4466 (4.71)      15.4466 (4.71)     0.0000 (1.0)       15.4466 (4.71)     0.0000 (1.0)           0;0  0.0647 (0.21)          1           1
test_taxi_trips_benchmark[8-False-V3]        25.9606 (7.91)      25.9606 (7.91)      25.9606 (7.91)     0.0000 (1.0)       25.9606 (7.91)     0.0000 (1.0)           0;0  0.0385 (0.13)          1           1
test_taxi_trips_benchmark[8-False-V2]        27.9552 (8.52)      27.9552 (8.52)      27.9552 (8.52)     0.0000 (1.0)       27.9552 (8.52)     0.0000 (1.0)           0;0  0.0358 (0.12)          1           1
test_taxi_trips_benchmark[8-True-V3]         28.0594 (8.55)      28.0594 (8.55)      28.0594 (8.55)     0.0000 (1.0)       28.0594 (8.55)     0.0000 (1.0)           0;0  0.0356 (0.12)          1           1
test_taxi_trips_benchmark[8-True-V2]         29.3288 (8.94)      29.3288 (8.94)      29.3288 (8.94)     0.0000 (1.0)       29.3288 (8.94)     0.0000 (1.0)           0;0  0.0341 (0.11)          1           1
test_taxi_trips_benchmark[16-False-V3]       52.4382 (15.98)     52.4382 (15.98)     52.4382 (15.98)    0.0000 (1.0)       52.4382 (15.98)    0.0000 (1.0)           0;0  0.0191 (0.06)          1           1
test_taxi_trips_benchmark[16-False-V2]       53.6786 (16.36)     53.6786 (16.36)     53.6786 (16.36)    0.0000 (1.0)       53.6786 (16.36)    0.0000 (1.0)           0;0  0.0186 (0.06)          1           1
test_taxi_trips_benchmark[16-True-V3]        55.9569 (17.05)     55.9569 (17.05)     55.9569 (17.05)    0.0000 (1.0)       55.9569 (17.05)    0.0000 (1.0)           0;0  0.0179 (0.06)          1           1
test_taxi_trips_benchmark[16-True-V2]        61.3706 (18.70)     61.3706 (18.70)     61.3706 (18.70)    0.0000 (1.0)       61.3706 (18.70)    0.0000 (1.0)           0;0  0.0163 (0.05)          1           1
test_taxi_trips_benchmark[100-False-V3]     323.4481 (98.56)    323.4481 (98.56)    323.4481 (98.56)    0.0000 (1.0)      323.4481 (98.56)    0.0000 (1.0)           0;0  0.0031 (0.01)          1           1
test_taxi_trips_benchmark[100-False-V2]     344.7229 (105.04)   344.7229 (105.04)   344.7229 (105.04)   0.0000 (1.0)      344.7229 (105.04)   0.0000 (1.0)           0;0  0.0029 (0.01)          1           1
test_taxi_trips_benchmark[100-True-V3]      391.4854 (119.29)   391.4854 (119.29)   391.4854 (119.29)   0.0000 (1.0)      391.4854 (119.29)   0.0000 (1.0)           0;0  0.0026 (0.01)          1           1
test_taxi_trips_benchmark[100-True-V2]      409.9862 (124.93)   409.9862 (124.93)   409.9862 (124.93)   0.0000 (1.0)      409.9862 (124.93)   0.0000 (1.0)           0;0  0.0024 (0.01)          1           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
24 passed in 1939.67s (0:32:19)

real    32m25.187s
user    3m11.710s
sys     0m15.668s
+ set +x
```

Note that results can be compared as follows:

```
---------------------------------------------------------- benchmark: 90 tests -----------------------------------------------------------
Name (time in s)                                                                                                             Min
------------------------------------------------------------------------------------------------------------------------------------------
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-True-V3] (initial_baseline_run_3)              3.2818 (1.0)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (initial_baseline_run_2)             3.3227 (1.01)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_initial_baseline_run_{3})       3.4572 (1.05)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_initial_baseline_run_{2})       3.5161 (1.07)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-True-V3] (initial_baseline_run_2)              3.5200 (1.07)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (initial_baseline_run_2)             3.5963 (1.10)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_initial_baseline_run_{i})       3.7672 (1.15)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (initial_baseline_run_3)             3.8316 (1.17)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_initial_baseline_run_{2})       3.8428 (1.17)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_initial_baseline_run_{i})       3.8469 (1.17)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (initial_baseline_run_3)             3.8848 (1.18)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_2_initial_baseline_run_2)       3.9233 (1.20)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-True-V2] (initial_baseline_run_2)              3.9647 (1.21)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_initial_baseline_run_2)         3.9726 (1.21)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_initial_baseline_run_1)         4.0638 (1.24)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_initial_baseline_run_{3})       4.1432 (1.26)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-True-V3] (initial_baseline_run_1)              4.2233 (1.29)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_2_initial_baseline_run_1)       4.2601 (1.30)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_2_initial_baseline_run_2)       4.2942 (1.31)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (initial_baseline_run_1)             4.3585 (1.33)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-True-V2] (initial_baseline_run_3)              4.4692 (1.36)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_initial_baseline_run_2)         4.4795 (1.36)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_2_initial_baseline_run_3)       4.4885 (1.37)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-True-V2] (initial_baseline_run_1)              4.7968 (1.46)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_initial_baseline_run_3)         4.8230 (1.47)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_initial_baseline_run_3)         4.8993 (1.49)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_2_initial_baseline_run_1)       5.0234 (1.53)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (tmp_initial_baseline_run_1)         5.1549 (1.57)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V2] (initial_baseline_run_1)             5.2558 (1.60)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[1-False-V3] (tmp_2_initial_baseline_run_3)       5.2586 (1.60)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-False-V3] (initial_baseline_run_2)             6.9249 (2.11)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-True-V3] (initial_baseline_run_2)              7.0319 (2.14)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-False-V2] (initial_baseline_run_2)             7.0574 (2.15)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-False-V2] (initial_baseline_run_3)             7.2650 (2.21)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-True-V2] (initial_baseline_run_2)              7.7079 (2.35)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-False-V3] (initial_baseline_run_3)             7.8152 (2.38)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-True-V2] (initial_baseline_run_3)              8.2846 (2.52)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-False-V3] (initial_baseline_run_1)             8.3849 (2.55)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-True-V3] (initial_baseline_run_1)              8.4396 (2.57)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-True-V3] (initial_baseline_run_3)              8.4899 (2.59)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-False-V2] (initial_baseline_run_1)             8.7447 (2.66)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[2-True-V2] (initial_baseline_run_1)              8.8481 (2.70)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-False-V3] (initial_baseline_run_3)            13.0151 (3.97)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-False-V3] (initial_baseline_run_2)            13.5262 (4.12)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-False-V2] (initial_baseline_run_2)            13.7150 (4.18)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-True-V3] (initial_baseline_run_3)             13.7633 (4.19)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-False-V2] (initial_baseline_run_3)            14.2186 (4.33)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-True-V3] (initial_baseline_run_2)             15.3856 (4.69)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-True-V3] (initial_baseline_run_1)             15.4317 (4.70)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-True-V2] (initial_baseline_run_3)             15.4466 (4.71)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-True-V2] (initial_baseline_run_1)             16.2528 (4.95)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-True-V2] (initial_baseline_run_2)             16.5584 (5.05)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-False-V3] (initial_baseline_run_1)            17.2043 (5.24)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[4-False-V2] (initial_baseline_run_1)            18.1744 (5.54)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-False-V3] (initial_baseline_run_3)            25.9606 (7.91)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-False-V2] (initial_baseline_run_3)            27.9552 (8.52)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-True-V3] (initial_baseline_run_3)             28.0594 (8.55)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-False-V3] (initial_baseline_run_2)            28.7145 (8.75)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-True-V3] (initial_baseline_run_1)             29.2475 (8.91)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-True-V2] (initial_baseline_run_3)             29.3288 (8.94)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-False-V2] (initial_baseline_run_2)            29.4374 (8.97)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-False-V3] (initial_baseline_run_1)            29.9335 (9.12)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-True-V2] (initial_baseline_run_2)             30.3670 (9.25)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-True-V3] (initial_baseline_run_2)             30.9665 (9.44)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-True-V2] (initial_baseline_run_1)             33.0360 (10.07)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[8-False-V2] (initial_baseline_run_1)            36.7383 (11.19)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-False-V3] (initial_baseline_run_3)           52.4382 (15.98)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-False-V2] (initial_baseline_run_3)           53.6786 (16.36)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-False-V3] (initial_baseline_run_2)           53.7829 (16.39)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-True-V3] (initial_baseline_run_3)            55.9569 (17.05)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-False-V2] (initial_baseline_run_2)           56.1152 (17.10)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-True-V2] (initial_baseline_run_2)            60.8217 (18.53)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-False-V3] (initial_baseline_run_1)           61.3593 (18.70)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-True-V2] (initial_baseline_run_3)            61.3706 (18.70)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-True-V3] (initial_baseline_run_1)            61.6513 (18.79)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-True-V2] (initial_baseline_run_1)            62.8486 (19.15)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-False-V2] (initial_baseline_run_1)           68.7432 (20.95)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[16-True-V3] (initial_baseline_run_2)            69.8583 (21.29)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-False-V3] (initial_baseline_run_3)         323.4481 (98.56)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-False-V3] (initial_baseline_run_2)         328.5219 (100.10)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-False-V2] (initial_baseline_run_3)         344.7229 (105.04)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-False-V2] (initial_baseline_run_2)         348.5321 (106.20)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-True-V3] (initial_baseline_run_2)          382.3974 (116.52)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-True-V2] (initial_baseline_run_2)          384.1241 (117.05)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-True-V2] (initial_baseline_run_1)          384.9857 (117.31)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-True-V3] (initial_baseline_run_3)          391.4854 (119.29)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-False-V3] (initial_baseline_run_1)         391.8299 (119.39)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-True-V2] (initial_baseline_run_3)          409.9862 (124.93)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-False-V2] (initial_baseline_run_1)         435.7602 (132.78)
tests/performance/test_bigquery_benchmarks.py::test_taxi_trips_benchmark[100-True-V3] (initial_baseline_run_1)          477.5333 (145.51)
------------------------------------------------------------------------------------------------------------------------------------------
```

jdimatteo committed:
Minimal multithreading implemented for V3

jdimatteo committed:
Minimal multithreading implemented for V2

jdimatteo committed:
unlimitted pool size

jdimatteo committed:
unlimited overflow

jdimatteo committed:
hacked in unlimitted python requests pool size

jdimatteo committed:
Add support to script for passing args to pytest

jdimatteo committed:
added fully unlimited run results

jdimatteo committed:
added test results for minimal implementation

jdimatteo committed:
added results for when unlimited overflow/pool in sqlalchemy but still limited on python requests pool size

jdimatteo committed:
added html runs

jdimatteo committed:
Made concurrency configurable and default to disabled.

jdimatteo committed:
Cleanup

jdimatteo committed:
V3 API now configured with unlimited threadpool and overflow to prevent http://sqlalche.me/e/14/3o7r

jdimatteo committed:
added more results, v2 6.3x faster with HTML and 15x without HTML

jdimatteo committed:
move patching from test script into async executor

jdimatteo committed:
Patch wasn't applied early enough so moved into method called by test util.

jdimatteo committed:
changelog

jdimatteo committed:
fix

jdimatteo committed:
cleanup

jdimatteo committed:
test fixes

jdimatteo committed:
cleanup

jdimatteo committed:
cleanup, including AsyncExecutor is now a context manager.

jdimatteo committed:
cleanup / docstrings

jdimatteo committed:
docstrings

jdimatteo committed:
remove intermediate results, add results from latest commit

jdimatteo commented on great_expectations/validation_operators/validation_operators.py:
```diff
@@ -304,41 +305,55 @@ def run(
-                batch_identifier = batch.active_batch_id
-            else:
-                batch_identifier = batch.batch_id
+        with AsyncExecutor(
```
This is where multithreading occurs using the V2 API.

For the V3 API, multiple threads are used in `Checkpoint.run`.  With the V3 API, `len(assets_to_validate)` is always 1 in this `ActionListValidationOperator.run` method. Providing  `max_workers=len(assets_to_validate)` ensures that no new threads are spawned here with the V3 API (since we already have multiple threads in `Checkpoint.run`).

jdimatteo replied:
Added comment block in the code explaining this.

jdimatteo commented on great_expectations/checkpoint/checkpoint.py:
```diff
@@ -266,50 +266,63 @@ def run(
-                    "expectation_suite_name"
-                )
-                action_list: list = substituted_validation_dict.get("action_list")
+        with AsyncExecutor(
```
This is where multithreading occurs using the V3 API.  The V2 API is not multithreaded here and depends on mulithreading added to [validation_operators.py](https://github.com/great-expectations/great_expectations/pull/3362/files#r703766570).

eap replied:
This comment would be useful in a short block comment above line 269. Maybe explain that by setting the max_workers with the length of validations, you are short-circuiting the multithreading logic (force concurrent execution) in the case of the v3 api's external call (I think that's how you described it to me).

jdimatteo replied:
Done. Please see added comment blocks here and in validation_operators.py.

jdimatteo committed:
empty commit to trigger tests to re-run

I can't reproduce the timeout failures shown at https://dev.azure.com/great-expectations/great_expectations/_build/results?buildId=16449&view=logs&j=1faf4978-6413-5e5c-0a0b-fea762862743&t=a0f9135e-66bf-5346-c544-c446e6a7170e&l=1449

jdimatteo committed:
Updated doc

jdimatteo committed:
cleanup comment

jdimatteo committed:
consolidate configuration into ConcurrencyConfig and update docstrings

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into feature/speedup_checkpoint_running_2

jdimatteo committed:
fix test failures introduced from upstream/develop merge

jdimatteo commented on great_expectations/core/async_executor.py:
```diff
@@ -0,0 +1,133 @@
+        return self
+
+    def __exit__(self, ext, exv, trb):
+        self.shutdown()
```
Feedback from Evan: add in docstring what is being done or not done with the exception, and if this is an OK outcome. Are these automatically propagated up?

jdimatteo replied:
Thanks @eap for the review feedback! This is fixed in latest commit a1897c7258f18ed318efea430a0bafb05422f5b3

jdimatteo committed:
Cleanup AsyncExecutor context manager implementation per @eap's review feedback -- thanks @eap!

jdimatteo committed:
Merge branch 'develop' into feature/speedup_checkpoint_running_2

jdimatteo committed:
remove accidentally added useless line

jdimatteo commented on great_expectations/core/async_executor.py:
```diff
@@ -81,7 +81,6 @@ def __exit__(self, exc_type, exc_value, traceback):
         self.shutdown()
         # Do not do use the context manager exception arguments, in order to get the desired default behavior where any
         # exceptions are raised here. More info at https://docs.python.org/3.9/reference/datamodel.html#object.__exit__.
-        AbstractContextManager.__exit__(self, exc_type, exc_value, traceback)
```
Wups, I didn't mean to include this line. This is removed in commit 367f14e.

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into feature/speedup_checkpoint_running_2

eap commented on great_expectations/core/async_executor.py:
```diff
@@ -0,0 +1,133 @@
+    WARNING: This class is experimental.
+    """
+
+    def __init__(self, future: Future = None, value: Any = None):
```
def \_\_init\_\_(self, future: __Optional__[Future] = None, value: Any = None):

jdimatteo replied:
Done

eap commented on great_expectations/core/async_executor.py:
```diff
@@ -0,0 +1,133 @@
+            AsyncResult(future=self._thread_pool_executor.submit(fn, *args, **kwargs))
+            if self._execute_concurrently
+            else AsyncResult(value=fn(*args, **kwargs))
+        )
```
Nit: This feels cludgy because it is using a multi-line ternary inside the return... why not just rework it to use standard conditional syntax?

```
if self._execute_concurrently:
   return AsyncResult(future=...
else:
    return AsyncResult(value=fn(...
```

jdimatteo replied:
Done

eap commented on great_expectations/core/async_executor.py:
```diff
@@ -0,0 +1,133 @@
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.shutdown()
+        # Do not do use the context manager exception arguments, in order to get the desired default behavior where any
```
desired behavior == does not catch or block exemptions?

jdimatteo replied:
Desired behavior is to not catch, not block, nor suppress any exceptions. I updated the comment to more explicitly state this.

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into feature/speedup_checkpoint_running_2

eap commented on tests/performance/test_bigquery_benchmarks.py:
```diff
@@ -101,12 +121,29 @@ def test_taxi_trips_benchmark(
+def _recursively_assert_actual_result_matches_expected_result_keys(
+    expected, actual, description_for_error_reporting
+):
+    """Only assert on keys that exist in the expected result, so that test is less fragile and doesn't incorrectly fail
```
Maybe note explicitly that the "expected" input can be a subset of actual, and that this validates the contents of the expected subset, not the entire result set.

jdimatteo replied:
Done.

eap reviewed (commented):
first pass

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into feature/speedup_checkpoint_running_2

jdimatteo committed:
cleanup per @eap's review feedback linked below -- thanks @eap!

https://github.com/great-expectations/great_expectations/pull/3362/#pullrequestreview-750705547

jdimatteo commented:
> first pass

Thanks for the review @eap! I addressed all your comments -- please take another look.

eap commented on great_expectations/datasource/sqlalchemy_datasource.py:
```diff
@@ -239,30 +240,38 @@ def __init__(
+                )
+                concurrency.add_sqlalchemy_create_engine_parameters(kwargs)
+
+                # if a connection string or url was provided, use that
```
Nit: I've noticed a few one-line comments missing punctuation (here, line 264 below this, and elsewhere in this PR). Can you give this a once over and clean it up? It stuck out to me here because it's worded in such a way that without a period it looks like a sentence got cut off.

jdimatteo replied:
Done

eap commented on great_expectations/data_context/types/base.py:
```diff
@@ -1063,6 +1063,55 @@ def make_notebooks_config(self, data, **kwargs):
+        # databases and/or be manually user configurable.
+        return 100
+
+    def add_sqlalchemy_create_engine_parameters(self, parameters):
```
type annotation? MutableMapping?

jdimatteo replied:
Done

eap reviewed (approved):
Looks good - at this point I've only got nitpicks. I like the overall direction and I appreciate you walking me through it the other day.

The only real frustrating issue involves the wrapped second call to concurrency, but that seems genuinely hard to get around as long as the v3 api wraps the v2 api. I think the in-line docs make it easy enough to understand at this point.

jdimatteo committed:
cleanup per @eap's review feedback linked below -- thanks @eap!

https://github.com/great-expectations/great_expectations/pull/3362/#pullrequestreview-751522358

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into feature/speedup_checkpoint_running_2

talagluck commented on great_expectations/data_context/types/base.py:
```diff
@@ -1063,6 +1063,57 @@ def make_notebooks_config(self, data, **kwargs):
+        bottlenecks.
+
+        Args:
+            parameters: SqlAlchemy create_engine parameters to add concurrency appropriate parameters to. If the
```
```suggestion
            parameters: SqlAlchemy create_engine parameters to which we add concurrency appropriate parameters. If the
```

jdimatteo replied:
Done

jdimatteo commented on great_expectations/data_context/types/base.py:
```diff
@@ -1063,6 +1063,57 @@ def make_notebooks_config(self, data, **kwargs):
+        # databases and/or be manually user configurable.
+        return 100
+
+    def add_sqlalchemy_create_engine_parameters(
```
add issue to document/allow how to pass through these in data context config

jdimatteo replied:
Added issue "Document how to pass through SqlAlchemy create_engine parameters" #3394

jdimatteo commented on great_expectations/data_context/types/base.py:
```diff
@@ -1063,6 +1063,57 @@ def make_notebooks_config(self, data, **kwargs):
+    @property
+    def max_database_query_concurrency(self) -> int:
+        """Max number of concurrent database queries to execute with mulithreading."""
+        # BigQuery has a limit of 100 for "Concurrent rate limit for interactive queries" as described at
```
follow on PR to user configure 100

talagluck replied:
just a reminder that when we do this, we should update the marshmallow schema as well.

jdimatteo commented:
Based on review feedback from @talagluck, I added an issue to track data docs speed improvement: https://github.com/great-expectations/great_expectations/issues/3396.

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into feature/speedup_checkpoint_running_2

talagluck commented on tests/data_context/test_concurrency_config.py:
```diff
@@ -0,0 +1,31 @@
+    DataContextConfig,
+    InMemoryStoreBackendDefaults,
+)
+
```
It might be trivial, but could we also add a test to demonstate that `concurrency` can be specified from the `great_expectations.yml`?

jdimatteo replied:
I added `test_concurrency_enabled_with_yaml` in latest commit below -- is this what you had in mind?

talagluck replied:
Hmm - after seeing this, I guess it's probably not necessary. I was picturing something more like an integration test that loaded from a full yaml config, but we test things like that elsewhere, so it's not needed here.

jdimatteo replied:
OK, I removed test_concurrency_enabled_with_yaml

talagluck commented on tests/performance/taxi_benchmark_util.py:
```diff
@@ -11,22 +11,25 @@
 
 def create_checkpoint(
-    number_of_tables: int, html_dir: Optional[str] = None
+    number_of_tables: int, backend_api: str = "V3", html_dir: Optional[str] = None
```
This is awesome!

talagluck reviewed (commented):
This looks great! Just a few minor nits, and I think we're ready to merge. Thanks, @jdimatteo!

jdimatteo committed:
@talagluck's commit suggestion

Co-authored-by: talagluck <tal@superconductive.com>

jdimatteo committed:
Added yaml test per @talagluck review feedback linked below

https://github.com/great-expectations/great_expectations/pull/3362#discussion_r708695456

jdimatteo committed:
Merge remote-tracking branch 'upstream/develop' into feature/speedup_checkpoint_running_2

jdimatteo committed:
Updated changelog to match PR title

jdimatteo commented:
> This looks great! Just a few minor nits, and I think we're ready to merge. Thanks, @jdimatteo!

Thanks for the review @talagluck! I addressed all your comments -- please take another look.

talagluck reviewed (approved):
Looks great, @jdimatteo - thanks for the updates!

I think the yaml test is probably not necessary, but I don't want to gate on that, and it doesn't hurt anything, so I'll approve for now. If you're able to remove it in the next couple of hours, great, and otherwise I'll merge this in.

Thanks for this awesome contribution!

jdimatteo committed:
Revert "Added yaml test per @talagluck review feedback linked below"

This reverts commit 88592ee09719066dc12adb9d96e804b414a7d52c.

jdimatteo committed:
Merge branch 'develop' into feature/speedup_checkpoint_running_2

jdimatteo commented:
> Looks great, @jdimatteo - thanks for the updates!
> 
> I think the yaml test is probably not necessary, but I don't want to gate on that, and it doesn't hurt anything, so I'll approve for now. If you're able to remove it in the next couple of hours, great, and otherwise I'll merge this in.
> 
> Thanks for this awesome contribution!

Thanks for the review @talagluck! I removed the yaml test in the latest commit and this is ready to merge from my perspective.
